tasks: add setup/teardown fixtures

osrjv opened the pull request:
Needs some documentation, but here's how it currently works:
```python
from robocorp.tasks import task, setup, teardown


@setup
def before_task(task):
    print(f"Running before task: {task.name}")


@setup(scope="session")
def before_all(tasks):
    names = ", ".join(task.name for task in tasks)
    print(f"Running before tasks: {names}")


@teardown
def after_task(task):
    print(f"Running after task: {task.name}")


@teardown(scope="session")
def after_all(task):
    names = ", ".join(task.name for task in tasks)
    print(f"Running after tasks: {names}")


@task
def first():
    print("Doing things")


@task
def second():
    print("Doing other things")
```

fabioz reviewed (commented):
+1 one from me. 

The one change I'd do is instead of using `all`, using `session` (since that's the same name already used in session_cache).

Also, I guess this needs some docs (especially to say that the setup/teardown must be explicitly imported to work).

cmin764 reviewed (commented):
LGTM! (including Fabio's comment)

fabioz commented:
I was thinking a bit more and I think that maybe the default should be `tasks`, not `session` if the argument isn't given (as I believe a setup per task is more common than a setup per session).

osrjv commented:
> I was thinking a bit more and I think that maybe the default should be `tasks`, not `session` if the argument isn't given (as I believe a setup per task is more common than a setup per session -- that also follows what `pytest` does, so, pytest users may be expecting that).

Makes sense. Would also make the argument of `ITask` instead of `Sequence[ITask]` a bit more obvious and easy to use. I think in 99% of cases we want users to only run one task at once anyways.

osrjv commented:
Review fixes:
- Changed naming to `task` and `session`
- Changed `task` to be default
- Added method docs and a guide

Will still change the setup callbacks to raise on error

fabioz commented:
I was actually thinking that most of my use-cases are actually a setup along with a teardown (so, writing a case where there's setup/yield/teardown would be nice).

osrjv commented:
> I was actually thinking that most of my use-cases are actually a setup along with a teardown (so, writing a case where there's setup/yield/teardown would be nice).

How do you see that from a usage and naming PoV? It's neither a setup nor a teardown, but more of a generic `fixture` like in pytest. I guess I could add an additional function that supports it, similar to the caching decorators

fabioz commented:
I think `fixture` is ok, but even `setup` accepting a `yield` inside it would work for me...

osrjv commented:
> I think `fixture` is ok, but even `setup` accepting a `yield` inside it would work for me...

I went with this. Here's what a yielding setup looks like:
```python
@setup
def measure_time(task):
    start = time.time()
    yield  # Task executes here
    duration = time.time() - start
    print(f"Task {task.name} took {duration} seconds")
```

Added also relevant documentation

cmin764 reviewed (commented):
Wondering what happens if you decorate with both a `setup` which yields and a `teardown`. Would both the below yielding part and the other teardown logic execute? (haven't tested it yet)

osrjv commented:
> Wondering what happens if you decorate with both a `setup` which yields and a `teardown`. Would both the below yielding part and the other teardown logic execute? (haven't tested it yet)

Yes, there is no limit to the amount of setup and teardown fixtures you can register. Setups are executed in registered order and teardowns in reverse registered order.

With this code:
```python
from robocorp.tasks import setup, teardown, task

@setup
def yielding(task):
    print("Generator before yield")
    yield
    print("Generator after yield")

@teardown
def first(task):
    print("Teardown #1")

@teardown
def second(task):
    print("Teardown #2")

@task
def a_task():
    print("Inside task")
```

The following would be printed:
```
Generator before yield
Inside task
Generator after yield
Teardown #2
Teardown #1
```

fabioz commented on tasks/docs/guides/02-setups-teardowns.md:
```diff
@@ -0,0 +1,101 @@
+@setup
+def before_and_after(task_before):
+    print(f"Task before running: {task_before}")
+    task_after = yield
```
This got me a bit puzzled. Isn't it so that both `task_before` and `task_after` are the same object?

i.e:  `assert task_before is task_after` should work.

If this is true, I'm not sure whether returning the task from yield does make much sense (not that it couldn't be done, but it's not very useful given that it'd just return the task in the parameter).

osrjv replied:
Good catch. I just noticed the same and had it already removed locally.

It would maybe be better if we didn't pass the mutable task object to the user, but maybe that has some use-cases.

osrjv replied:
We might have to change the inner workings a bit to actually support passing a run by changing the status of the task objects, I think? I'll try it out, actually

fabioz reviewed (approved):
LGTM!
